### PR TITLE
[SandboxIR] PassManager

### PR DESCRIPTION
**Pull request by vporpo**
This patch implements a simple pass manager for Sandbox IR.

### CODE REVIEW

**Review by Sterling-Augustine — APPROVED** [review]
All pretty standard.

File: llvm/include/llvm/SandboxIR/PassManager.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    // TODO: Check that Pass's class type works with this PassManager type.
+    Passes.push_back(Pass);
+  }
+#ifndef NDEBUG
+  void print(raw_ostream &OS) const override {
+    OS << this->getName();
+    OS << "(";
+    for (auto [Idx, Pass] : enumerate(Passes)) {
+      OS << Pass->getName();
+      if (Idx + 1 != Passes.size())
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/PassManager.h:53]: can this use `interleaveComma`?

File: llvm/include/llvm/SandboxIR/PassManager.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Registers and executes the Sandbox IR passes.
+//
+// The pass manager contains an ordered sequence of passes that it runs in
```
Comment (aeubanks) [inline llvm/include/llvm/SandboxIR/PassManager.h:11]: it seems error prone for the pass manager to not own the pass (e.g. stack references). e.g. in the new pass manager if you want to run a pass multiple times, you just create multiple instances of it

Comment (vporpo) [reply llvm/include/llvm/SandboxIR/PassManager.h:11]: The pass registry would own the passes and would also provide the name-to-pass lookup so that you can build the pipeline from a pipeline string like "pass1,pass2".

If the pass manager owns the passes, then how is the mapping from names to passes handled?

File: llvm/include/llvm/SandboxIR/PassManager.h (hunk 1 above)
Comment (vporpo) [reply llvm/include/llvm/SandboxIR/PassManager.h]: `interleaveComma` inserts a whitespace after the comma, so I just used `interleave()`

File: llvm/include/llvm/SandboxIR/PassManager.h (hunk 2 above)
Comment (aeubanks) [reply llvm/include/llvm/SandboxIR/PassManager.h:11]: see includes of `PassRegistry.def`, especially in `PassBuilder.cpp`. it's basically a string comparison chain

I think your way is fine, as long as your explanation of the registry owning passes is documented

**Review by aeubanks — APPROVED** [review]

File: llvm/include/llvm/SandboxIR/PassManager.h (hunk 2 above)
Comment (vporpo) [reply llvm/include/llvm/SandboxIR/PassManager.h:11]: I see, it shouldn't be too hard to follow a similar design. Let me try it out.

Comment (vporpo) [reply llvm/include/llvm/SandboxIR/PassManager.h:11]: I think I will stick with this design for now. Using a PassRegistry.def complicates the design quite a bit and I am not sure it is worth it at this point. We can revisit it later if needed.
